web/controllers/api/docs_controller.ex

```diff
@@ -163,16 +163,13 @@ defmodule HexWeb.API.DocsController do
     HexWeb.CDN.purge_key(:fastly_hexdocs, unversioned_key)
     HexWeb.CDN.purge_key(:fastly_hexdocs, versioned_key)
 
-    # Set docs flag on release
-    HexWeb.Repo.transaction(fn ->
-      Ecto.Changeset.change(release, has_docs: true)
-      |> HexWeb.Repo.update!
+    multi =
+      Ecto.Multi.new
+      |> Ecto.Multi.update(:release, Ecto.Changeset.change(release, has_docs: true))
+      |> Ecto.Multi.update(:package, Ecto.Changeset.change(release.package, docs_updated_at: Ecto.DateTime.utc))
+      |> Ecto.Multi.run(:log, fn _ -> audit(user, "docs.publish", {package, release}) end)
```

Comment:
@josevalim thanks for the feedback, I really appreciate it.

I tried replacing the code above (Ecto.Multi etc) with the following code:

   ```elixir
    changeset =
      release
      |> Ecto.Changeset.change(has_docs: true)
      |> Ecto.Changeset.put_assoc(:package, Ecto.Changeset.change(release.package, docs_updated_at: Ecto.DateTime.utc))
      |> Ecto.Changeset.put_assoc(:audit_log, %AuditLog{action: "docs.publish", actor: user, params: %{
        package: %{name: package.name, id: package.id},
        release: %{}, # ???
      }})
    HexWeb.Repo.update!(changeset)
   ```

I finally can see the advantages of using assocs vs Ecto.Multi when the latter is not absolutely necessary - we also haven't lost readability which Multi provides.

I ran into an issue though, I got the following error:

```** (RuntimeError) attempting to cast or change association `audit_log` from `HexWeb.Release` that was not loaded```

I fixed it, by doing this:

```elixir
release = ... # fetch release from DB
release = Map.put(release, :audit_log, nil)
```

which feels kinda dirty. Any ideas of how to improve it? I suppose this line could be moved to each models `changeset` function so at least it wouldn't be spread all over the app.

As you can see I can easily grab `package.id` for this particular AuditLog, but I also want to have `release.id` which is not available until changeset is committed to the DB. You mentioned the fake `logger_id` field - could you expand on that, how exactly would it work in this particular case? Fake as in virtual field, or would it be the actual DB column? Would there be some custom `AuditLog.changeset/2` that `put_assoc` would automagically call or something like that?